src/charm.py

```diff
@@ -361,6 +374,53 @@ def set_main_unit(self, unit: str) -> None:
             logging.info("Setting main unit to be %s", unit)
             peer_relation[0].data[self.app].update({MAIN_UNIT_ID: unit})
 
+    def set_signing_key(self, signing_key: str) -> None:
+        """Create secret with signing key content.
+
+        Args:
+            signing_key: signing key as string.
+        """
+        peer_relation = self.model.relations[synapse.SYNAPSE_PEER_RELATION_NAME]
+        if not peer_relation:
+            logger.error(
+                "Failed to set signing key: no peer relation %s found",
+                synapse.SYNAPSE_PEER_RELATION_NAME,
+            )
+            return
+
+        if signing_key == self.get_signing_key():
+            logger.info("Received signing key but there is no change, skipping")
+            return
+        logger.debug("Adding signing key to secret: %s", signing_key)
+        secret = self.app.add_secret({"secret-signing-key": signing_key})
+        peer_relation[0].data[self.app].update({"secret-signing-id": typing.cast(str, secret.id)})
+
+    def get_signing_key(self) -> typing.Optional[str]:
+        """Get signing key from secret.
+
+        Returns:
+            Signing key as string or None if not found.
+        """
+        peer_relation = self.model.relations[synapse.SYNAPSE_PEER_RELATION_NAME]
+        if not peer_relation:
+            logger.error(
```

Comment:
I wouldn't log this if only a single instance is deployed. Also, shouldn't an exception be raised when there're no peer relations but there're multiple units?

Reply:
A synapse_peers_relation_created event is emitted after the start and the relation synapse-peers is created even for 1 unit. This check is more for a really exceptional case (considering peers) and also to make lint happy.

Currently, we return None if a relation is not found in the charm but this could be a subject for a refactoring and also a point to the Contributing guide.

Reply:
Gotcha